### PR TITLE
IC-1356 allow end endpoint to accept cancellation reason

### PR DESCRIPTION
**Pull request by rgforsyth**
## What does this pull request do?

Allows a cancellation reason to be passed the the 'end' endpoint.

## What is the intent behind these changes?

Allows a cancellation reason to be persisted when ending a referral.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/mappers/CancellationReasonMapper.kt [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.entity.CancellationReason
+import uk.gov.justice.digital.hmpps.hmppsinterventionsservice.jpa.repository.CancellationReasonRepository
+
+@Component
+class CancellationReasonMapper(
+  private val cancellationReasonRepository: CancellationReasonRepository
+) {
+  fun mapCancellationReasonIdToCancellationReason(cancellationReasonId: String): CancellationReason {
+    return cancellationReasonRepository.findByIdOrNull(cancellationReasonId)
+      ?: throw ResponseStatusException(HttpStatus.NOT_FOUND, "Cancellation reason not found [id=$cancellationReasonId]")
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/mappers/CancellationReasonMapper.kt:16]: i'm not sure 404 is the appropriate response here. if we raised `EntityNotFound` we would also return 404. but i think bad request might be better. i.e. the cancellation code is invalid

Comment (tomsmyers) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/controller/mappers/CancellationReasonMapper.kt:16]: or have this method raise entitynotfound, and have the controller catch it and return bad request

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/EndedReferralDTO.kt [hunk 2]
```diff
@@ -21,6 +22,7 @@ class EndedReferralDTO(
         referenceNumber = referral.referenceNumber!!,
         assignedTo = referral.assignedTo?.let { AuthUserDTO.from(it) },
         referral = DraftReferralDTO.from(referral),
+        cancellationReason = referral.cancellationReason!!.id
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/EndedReferralDTO.kt:25]: is the id the best thing to include in this response? i feel like the string representation would be more useful.

**Review by tomsmyers — COMMENTED** [review]
couple small things

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/EndedReferralDTO.kt (hunk 2 above)
Comment (rgforsyth) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/EndedReferralDTO.kt:25]: Just to clarify. The 'id' in this case is the 3 letter string code. ('MIS') for example. The alternative would be providing the description, or both?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/EndedReferralDTO.kt [hunk 3]
```diff
@@ -21,6 +24,9 @@ class EndedReferralDTO(
         referenceNumber = referral.referenceNumber!!,
         assignedTo = referral.assignedTo?.let { AuthUserDTO.from(it) },
         referral = DraftReferralDTO.from(referral),
+        cancellationReason = referral.cancellationReason!!.description,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/EndedReferralDTO.kt:27]: please can we leave these as nullable for now? i'm not sure that ended referrals will always have cancellation reasons and comments

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/EndReferralDTO.kt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+package uk.gov.justice.digital.hmpps.hmppsinterventionsservice.dto
+
+data class EndReferralDTO(
+  val cancellationReasonCode: String,
+  val cancellationComments: String,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/dto/EndReferralDTO.kt:5]: please can we make this optional?

**Review by tomsmyers — APPROVED** [review]
awesome- thanks

**Pull request merged**